src/nf_heuristic.py

```diff
@@ -0,0 +1,570 @@
+from __future__ import print_function
+import os
+import subprocess
+import collections
+import argparse
+import itertools
+import nfcp_user_level_parser as configParser
+import nfcp_library_parser as libParser
+import nfcp_code_generator as codeGenerator
+import nfcp_bess_generator as bessGenerator
+import util.lang_parser_helper as lang_helper
+import new_bess_generator as BG
+from termcolor import colored
+from util.nfcp_nf_node import *
+import util.nfcp_nf_node as ND
+from util.nfcp_install_table_entry import NFCP_entry_helper
+from core.profile_p4 import p4_usage_checker
+from connect import stage_feasible
+from nfcp_compiler import get_argparse
+from user_level_parser.UDNFCPUserListener \
+    import convert_nf_graph, convert_global_nf_graph
+import nf_placement as placeTool
+from nf_placement import log_module
+import heuristic_util.graph as G
+
+STATE_INIT = 0
+global CONF_LIB
+OUTPUT_DIR = "./"
+global P414_LIB_DIR
+CONF_LIB_DIR = "./user_level_examples"
+P414_LIB_DIR = "./p4_14_lib/"
+option_nf = []
+global last_index
+last_index = -1
+global option_set
+option_set = []
+
+def examine_prev(module_list):
+    for i in range(len(module_list)-1):
+        print("module %s: its previous is " % module_list[i+1].nf_class)
+        for prev in module_list[i+1].prev_nodes:
+            print("%s "% prev.nf_class)
+
+def parse_all_node(nfcp_parser):
+    nf_graph = convert_global_nf_graph(nfcp_parser.scanner)
+    modules = nf_graph.list_modules()
+    modules.sort(key=lambda l: (l.service_path_id, l.service_id))
+    return modules
+
+def init_allp4(module_list):
```

Comment:
In my impression, the following several functions are critical. It is helpful if you could add 1-2 sentenses saying what it does. Also, remember to remove unnecessary comments from the code.